### PR TITLE
towards complex-valued problems (second try)

### PR DESCRIPTION
**Pull request by davydden**
this is pretty much everything "complex" branch currently has but FE-values related changes, which are already merged in https://github.com/dealii/dealii/pull/918. The only useful stuff left there are complex petsc/slepc unit tests. 

The biggest change here is related to constraint matrix. I temporary comment-out some instantiation in MatrixTools until the pull-request is in a good shape.

p.s. first attempt with whole lot of changes was here https://github.com/dealii/dealii/pull/472

### CODE REVIEW

File: include/deal.II/lac/petsc_vector_base.h [hunk 1]
```diff
@@ -848,8 +863,10 @@ namespace PETScWrappers
 
       const PetscInt petsc_i = index;
 
+      PetscScalar cast_value = static_cast<PetscScalar> (value);
+
       const int ierr
-        = VecSetValues (vector, 1, &petsc_i, &value, INSERT_VALUES);
+        = VecSetValues (vector, 1, &petsc_i, &cast_value, INSERT_VALUES);
```
Comment (bangerth) [inline include/deal.II/lac/petsc_vector_base.h]: I don't think I understand this. `value` has type `const PetscScalar &`. `cast_value` has type `PetscScalar`. If you take the address, you would get `const PetscScalar *` and `PetscScalar *`, respectively. So the only difference is in the `const`. But `VecSetValues` takes a const pointer as fourth argument, as per http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Vec/VecSetValues.html . So it seems to me that you don't really need this change. Any insight?

Comment (bangerth) [inline include/deal.II/lac/petsc_vector_base.h]: As a side remark, there is also `VecSetValues` that is meant for the addition of a single value and that could be used here: http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Vec/VecSetValue.html#VecSetValue . But this is for a separate patch.

File: include/deal.II/lac/petsc_vector_base.h [hunk 2]
```diff
@@ -1003,10 +1020,37 @@ namespace PETScWrappers
...
+
+
+    inline
+    PetscReal
+    VectorReference::imag () const
+    {
+#ifndef PETSC_USE_COMPLEX
+      // This is a no op if complex numbers are not defined.
+      AssertThrow (false, ExcNotImplemented ());
+      return PetscReal (0);
```
Comment (bangerth) [inline include/deal.II/lac/petsc_vector_base.h]: I would remove the assertion. If you're not using complex scalars, then their imaginary parts are simply zero.

File: source/distributed/solution_transfer.cc [hunk 3]
```diff
@@ -216,18 +216,18 @@ namespace parallel
...
       const unsigned int dofs_per_cell=cell->get_fe().dofs_per_cell;
-      ::dealii::Vector<double> dofvalues(dofs_per_cell);
+      ::dealii::Vector<typename VECTOR::value_type> dofvalues(dofs_per_cell);
       for (typename std::vector<const VECTOR *>::iterator it=input_vectors.begin();
            it !=input_vectors.end();
            ++it)
         {
           cell->get_interpolated_dof_values(*(*it), dofvalues);
-          std::memcpy(data_store, &dofvalues(0), sizeof(double)*dofs_per_cell);
+          std::memcpy(data_store, &dofvalues(0), sizeof(typename VECTOR::value_type)*dofs_per_cell);
```
Comment (bangerth) [inline source/distributed/solution_transfer.cc]: I think you will also need to adjust the size of the `data_store` array for this. If you don't want to do this change, please at least add an
```
  Assert (typeid(typename VECTOR::value_type) == typeid(double), ExcNotImplemented());
```
here.

File: source/distributed/solution_transfer.cc [hunk 4]
```diff
@@ -245,14 +245,14 @@ namespace parallel
...
-      const double *data_store = reinterpret_cast<const double *>(data);
+      ::dealii::Vector<typename VECTOR::value_type> dofvalues(dofs_per_cell);
+      const typename VECTOR::value_type *data_store = reinterpret_cast<const typename VECTOR::value_type *>(data);
 
       for (typename std::vector<VECTOR *>::iterator it = all_out.begin();
            it != all_out.end();
            ++it)
         {
-          std::memcpy(&dofvalues(0), data_store, sizeof(double)*dofs_per_cell);
+          std::memcpy(&dofvalues(0), data_store, sizeof(typename VECTOR::value_type)*dofs_per_cell);
```
Comment (bangerth) [inline source/distributed/solution_transfer.cc]: same here.

File: source/lac/full_matrix.cc [hunk 5]
```diff
@@ -20,6 +20,12 @@ DEAL_II_NAMESPACE_OPEN
 
 #include "full_matrix.inst"
 
+// This is needed if PETSc was compiled with complex, though, it may
+// be used elsewhere too.
+//#ifdef PETSC_USE_COMPLEX
+template void dealii::FullMatrix<double>::vmult<std::complex<double> >(dealii::Vector<std::complex<double> > &, dealii::Vector<std::complex<double> > const &, bool) const;
+//#endif // PETSC_USE_COMPLEX
```
Comment (bangerth) [inline source/lac/full_matrix.cc]: If you don't need th `// #ifdef`, just remove it.

File: source/lac/petsc_parallel_sparse_matrix.cc [hunk 6]
```diff
@@ -858,7 +858,8 @@ namespace PETScWrappers
     {
       Vector tmp (v);
       vmult (tmp, v);
-      return tmp*v;
+      //note, that v*tmp returns conjugate(v) * tmp
```
Comment (bangerth) [inline source/lac/petsc_parallel_sparse_matrix.cc]: that's confusing. I think what you're saying is something like this:
```
  //note, that v*tmp returns  sum_i conjugate(v)_i * tmp_i
```

File: source/lac/petsc_parallel_sparse_matrix.cc [hunk 7]
```diff
@@ -867,6 +868,7 @@ namespace PETScWrappers
     {
       Vector tmp (v);
       vmult (tmp, v);
+      //note, that u*tmp returns conjugate(u) * tmp
```
Comment (bangerth) [inline source/lac/petsc_parallel_sparse_matrix.cc]: same here

File: source/numerics/matrix_tools.inst.in [hunk 8]
```diff
@@ -54,40 +54,40 @@ for (deal_II_dimension : DIMENSIONS; deal_II_space_dimension :  SPACE_DIMENSIONS
...
-      void MatrixCreator::create_mass_matrix<deal_II_dimension,float,deal_II_space_dimension>
-      (const DoFHandler<deal_II_dimension,deal_II_space_dimension>    &dof,
-       const Quadrature<deal_II_dimension>    &q,
-       SparseMatrix<float>     &matrix,
-       const Function<deal_II_space_dimension>      &rhs,
-       Vector<double>           &rhs_vector,
-       const Function<deal_II_space_dimension> * const coefficient,
-       const ConstraintMatrix   &constraints);
+    //template
+    //  void MatrixCreator::create_mass_matrix<deal_II_dimension,float,deal_II_space_dimension>
```
Comment (bangerth) [inline source/numerics/matrix_tools.inst.in]: if you don't need these instantiations, just remove them altogether. no need to leave them in as comments.

File: source/numerics/matrix_tools.inst.in [hunk 9]
```diff
@@ -166,26 +166,26 @@ for (deal_II_dimension : DIMENSIONS; deal_II_space_dimension :  SPACE_DIMENSIONS
...
-      (const hp::MappingCollection<deal_II_dimension,deal_II_space_dimension>       &mapping,
-       const hp::DoFHandler<deal_II_dimension,deal_II_space_dimension>    &dof,
-       const hp::QCollection<deal_II_dimension>    &q,
-       SparseMatrix<float>     &matrix,
-       const Function<deal_II_space_dimension>      &rhs,
-       Vector<double>           &rhs_vector,
-       const Function<deal_II_space_dimension> * const coefficient,
-       const ConstraintMatrix   &constraints);
+    //template
+    //  void MatrixCreator::create_mass_matrix<deal_II_dimension>
```
Comment (bangerth) [inline source/numerics/matrix_tools.inst.in]: same here and below

File: source/distributed/solution_transfer.cc (hunk 3 above)
Comment (davydden) [inline source/distributed/solution_transfer.cc]: are u sure? data_store is just a pointer to the start of the vector. It should have the same size irrespectively of type, as far as i understand.

**Comment by bangerth** [comment]
That turned out to be less terrible than I thought (thought it must have been terribly boring to write...). Can you address the few comments, squash into a set of self-contained commits (not necessarily just one commit, but for example one per independent set of changed), and upload again?

**Comment by bangerth** [comment]
Also, thanks for working on this!

File: source/numerics/matrix_tools.inst.in (hunk 8 above)
Comment (davydden) [inline source/numerics/matrix_tools.inst.in]: the point is, that currently matrix creator will not compile with those as it tries to use matrix with arbitrary base type (e.g. `float`) together with `Vector<double>`. That does not play well with my modifications to constraint matrix. I am ok to remove those if nobody needs those float instantiation. Ok keep it commented out until MatrixCreator class is adjusted accordingly

File: source/distributed/solution_transfer.cc (hunk 3 above)
Comment (bangerth) [inline source/distributed/solution_transfer.cc]: What I meant is that the array to which data_store points needs to be sized correctly :-)

File: source/numerics/matrix_tools.inst.in (hunk 8 above)
Comment (bangerth) [inline source/numerics/matrix_tools.inst.in]: Nah, let's force people to using matching scalar types for the local and global matrices.

File: source/distributed/solution_transfer.cc (hunk 3 above)
Comment (davydden) [inline source/distributed/solution_transfer.cc]: sure ;-) but it's sized somewhere outside of `pack_callback`/`unpack_callback`. But I must admit I don't understand in full the execution logic and what ends up in that void pointer.

Comment (bangerth) [inline source/distributed/solution_transfer.cc]: Right. That's why I suggested the assertion, so that someone else can deal with it later ;-)

**Comment by davydden** [comment]
ok, will squash later during the week...

**Comment by bangerth** [comment]
Good plan -- let me know when you're ready!

**Comment by davydden** [comment]
@bangerth i stashed. You might want to consider running testsuite before merging, I am on OSX and tests are not robust here yet...

**Comment by bangerth** [comment]
Will do.

I have this warning you may want to address in your commits:
```
/u/bangerth/p/deal.II/1/dealii/source/fe/fe_tools_interpolate.cc:225:48: warning: unused variable âtouch_iâ [-Wunused-variable]
```
Tests will take a while, will report back here later.

**Comment by bangerth** [comment]
Here's another set of failures, from `tests/matrix_free/step-37.debug`
```
5553: /u/bangerth/p/deal.II/1/dealii/include/deal.II/lac/constraint_matrix.h: In member function ‘void dealii::ConstraintMatrix::distribute_local_to_global(const InVector&, const std::vector<unsigned int>&, OutVector&) const [with InVector = dealii::FullMatrix<double>, OutVector = dealii::FullMatrix<float>]’:
5553: /u/bangerth/p/deal.II/1/dealii/tests/matrix_free/step-37.cc:555:13:   instantiated from ‘void Step37::LaplaceProblem<dim>::assemble_multigrid() [with int dim = 2]’
5553: /u/bangerth/p/deal.II/1/dealii/tests/matrix_free/step-37.cc:628:9:   instantiated from ‘void Step37::LaplaceProblem<dim>::run() [with int dim = 2]’
5553: /u/bangerth/p/deal.II/1/dealii/tests/matrix_free/step-37.cc:648:25:   instantiated from here
5553: /u/bangerth/p/deal.II/1/dealii/include/deal.II/lac/constraint_matrix.h:1602:3: error: no match for ‘operator==’ in ‘(& local_vector)->dealii::FullMatrix<double>::<anonymous>.dealii::Table<2, double>::<anonymous>.dealii::TableBase<N, T>::size [with int N = 2, T = double]() == (& local_dof_indices)->std::vector<_Tp, _Alloc>::size [with _Tp = unsigned int, _Alloc = std::allocator<unsigned int>, std::vector<_Tp, _Alloc>::size_type = long unsigned int]()’
5553: /u/bangerth/p/deal.II/1/dealii/include/deal.II/lac/constraint_matrix.h:1602:3: note: candidates are:
5553: /u/bangerth/p/deal.II/1/dealii/include/deal.II/base/table_indices.h:56:8: note: bool dealii::TableIndicesBase<N>::operator==(const dealii::TableIndicesBase<N>&) const [with int N = 2]
5553: /u/bangerth/p/deal.II/1/dealii/include/deal.II/base/table_indices.h:56:8: note:   no known conversion for argument 1 from ‘std::vector<unsigned int>::size_type {aka long unsigned int}’ to ‘const dealii::TableIndicesBase<2>&’
5553: /u/bangerth/p/deal.II/1/dealii/include/deal.II/base/aligned_vector.h:808:6: note: template<class T> bool dealii::operator==(const dealii::AlignedVector<T>&, const dealii::AlignedVector<T>&)
5553: /u/bangerth/p/deal.II/1/dealii/include/deal.II/lac/constraint_matrix.h:1602:3: error: no matching function for call to ‘dealii::StandardExceptions::ExcDimensionMismatch::ExcDimensionMismatch(const dealii::TableIndices<2>&, std::vector<unsigned int>::size_type)’
```
There are many more errors after this for this one test.

My tests aren't going to finish before bedtime. If you want to address these already, feel free and I'll try again tomorrow morning.

**Comment by bangerth** [comment]
So far, the tests that failed are these:
```
 6/7395 Test #5356: lac/vector_view_03.release .........................................***Failed   38.69 sec
  7/7395 Test #5220: lac/sparsity_pattern_10.debug ......................................***Failed   38.97 sec
  9/7395 Test #5219: lac/identity_matrix_05.release .....................................***Failed   38.98 sec
 10/7395 Test #5221: lac/sparsity_pattern_10.release ....................................***Failed   38.97 sec
 16/7395 Test #6583: multigrid/transfer_block_select.release ............................***Failed   16.49 sec
 17/7395 Test #6586: multigrid/transfer_system_adaptive_02.debug ........................***Failed    1.81 sec
 18/7395 Test #6587: multigrid/transfer_system_adaptive_02.release ......................***Failed    2.21 sec
 19/7395 Test #6588: multigrid/transfer_block.debug .....................................***Failed    0.04 sec
 20/7395 Test #6589: multigrid/transfer_block.release ...................................***Failed    1.74 sec
 28/7395 Test #5895: mpi/fe_field_function_01.mpirun=10.release .........................***Failed   72.13 sec
 29/7395 Test #5900: mpi/ghost_02.mpirun=10.debug .......................................***Failed   72.33 sec
 30/7395 Test #5948: mpi/renumber_cuthill_mckee.mpirun=1.debug ..........................***Failed   72.43 sec
248/7395 Test #2331: bits/parameter_handler_1_include_fail.debug ........................***Failed  161.12 sec
251/7395 Test #2091: bits/anna_3.release ................................................***Failed  161.96 sec
252/7395 Test #2086: bits/vector_44.release .............................................***Failed  162.56 sec
253/7395 Test #1963: bits/fe_field_function_01.release ..................................***Failed  162.56 sec
258/7395 Test #1638: bits/dof_constraints_02.debug ......................................***Failed  155.03 sec
344/7395 Test #7372: umfpack/umfpack_07.debug ...........................................***Failed   24.26 sec
738/7395 Test #870: base/index_set_10.debug ............................................***Failed    2.02 sec
2803/7395 Test #7373: umfpack/umfpack_07.release .........................................***Failed   14.33 sec
4787/7395 Test #4789: hp/hp_hanging_nodes_02.release .....................................***Failed    0.72 sec
5391/7395 Test #7374: umfpack/umfpack_01.debug ...........................................***Failed    1.82 sec
6025/7395 Test #5554: matrix_free/step-37.release ........................................***Failed    9.45 sec
6029/7395 Test #5553: matrix_free/step-37.debug ..........................................***Failed   11.12 sec
```
I run these with `ctest -j64` and a subset always fails for unrelated reasons. Not all of these may therefore be real failures.

**Comment by davydden** [comment]
Step-37 errors (https://github.com/dealii/dealii/pull/1008#issuecomment-112628019) are related to calling distribute_local_to_global for mismatch types:
```
InVector = dealii::FullMatrix<double>, OutVector = dealii::FullMatrix<float>]
```
At least part of those failing tests could be related to the choices made for the ConstraintMatrix in this patch. How shall we proceed?

**Comment by bangerth** [comment]
The tests that really are failing are these:
```
        5553 - matrix_free/step-37.debug (Failed)
        5554 - matrix_free/step-37.release (Failed)
        7372 - umfpack/umfpack_07.debug (Failed)
        7373 - umfpack/umfpack_07.release (Failed)
```
I'll take a look at them.

**Comment by davydden** [comment]
thanks!

**Comment by bangerth** [comment]
I committed all of the patches of this PR as part of #1027. Thanks, @davydden , for all of the work!

**Comment by davydden** [comment]
great!

**Comment by davydden** [comment]
in reference to https://github.com/dealii/dealii/issues/2033

**Pull request closed without merging**